apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsMessageListenerInstrumentation.java

```diff
@@ -0,0 +1,140 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.jms;
+
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.transaction.TraceContext;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.Queue;
+import javax.jms.Topic;
+
+import static co.elastic.apm.agent.jms.JmsInstrumentationHelper.JMS_TRACE_PARENT_HEADER;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+public class JmsMessageListenerInstrumentation extends BaseJmsInstrumentation {
+
+    @VisibleForAdvice
+    @SuppressWarnings("WeakerAccess")
+    public static final Logger logger = LoggerFactory.getLogger(JmsMessageListenerInstrumentation.class);
+
+    public JmsMessageListenerInstrumentation(ElasticApmTracer tracer) {
+        super(tracer);
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface()).and(hasSuperType(named("javax.jms.MessageListener")));
```

Comment:
We can't do pre-filtering for these as it might be implemented as an anonymous inner class or lambda. How big is the impact in terms of startup performance?

Reply:
Yeah, that's a given. There is an impact:
<img width="676" alt="Screen Shot 2019-06-06 at 10 54 13" src="https://user-images.githubusercontent.com/41850454/59016260-85c85680-8849-11e9-90e3-86ff45cd3818.png">
Hard to tell what real-world scenarios implications will be...
We do filter by ClassLoader, but this wouldn't help in any case JMS APIs are added in the app server level

Reply:
We decided to go with that, as it is.
Options discussed for relieving startup time overhead:
1. using `application_packages` filtering on type match- will capture all user's implementations of `MessageListener`s, but will miss framework listeners, like Message beans
2. add a mechanism for lazily match and instrument (`retransform`) already loaded classes on a background thread using some non-greedy strategy, forcing only specific instrumentations to be applied on runtime. For example- `MessageConsumer#setMessageListener` should be instrumented on startup because a consumer could be long-lived and we don't want to miss first invocation.

We can reconsider in the future. So far, we didn't hear any complaints about startup overhead/delays.